ArduCopter/RC_Channel_Copter.cpp

```diff
@@ -706,68 +712,78 @@ void RC_Channel_Copter::do_aux_function_change_force_flying(const AuxSwitchPos c
     }
 }
 
+// note that this is a method on the RC_Channels object, not the
+// individual channel
 // save_trim - adds roll and pitch trims from the radio to ahrs
-void Copter::save_trim()
+void RC_Channels_Copter::save_trim()
 {
     // save roll and pitch trim
-    float roll_trim = ToRad((float)channel_roll->get_control_in()*0.01f);
-    float pitch_trim = ToRad((float)channel_pitch->get_control_in()*0.01f);
-    ahrs.add_trim(roll_trim, pitch_trim);
+    float roll_trim = ToRad((float)get_roll_channel().get_control_in()*0.01f);
+    float pitch_trim = ToRad((float)get_pitch_channel().get_control_in()*0.01f);
+    AP::ahrs().add_trim(roll_trim, pitch_trim);
     LOGGER_WRITE_EVENT(LogEvent::SAVE_TRIM);
     gcs().send_text(MAV_SEVERITY_INFO, "Trim saved");
 }
 
+#if AP_COPTER_AHRS_AUTO_TRIM_ENABLED
+// start/stop ahrs auto trim
+void RC_Channels_Copter::do_aux_function_ahrs_auto_trim(const RC_Channel::AuxSwitchPos ch_flag)
+{
+    switch (ch_flag) {
+    case RC_Channel::AuxSwitchPos::HIGH:
+        gcs().send_text(MAV_SEVERITY_INFO, "AutoTrim running");
+        // flash the leds
+        AP_Notify::flags.save_trim = true;
+        auto_trim.running = true;
+        break;
+    case RC_Channel::AuxSwitchPos::MIDDLE:
+        break;
+    case RC_Channel::AuxSwitchPos::LOW:
+        if (auto_trim.running) {
+            AP_Notify::flags.save_trim = false;
+            save_trim();
+        }
+        break;
+    }
+}
+
 // auto_trim - slightly adjusts the ahrs.roll_trim and ahrs.pitch_trim towards the current stick positions
 // meant to be called continuously while the pilot attempts to keep the copter level
-void Copter::auto_trim_cancel()
+void RC_Channels_Copter::auto_trim_cancel()
 {
-    auto_trim_counter = 0;
+    auto_trim.running = false;
     AP_Notify::flags.save_trim = false;
     gcs().send_text(MAV_SEVERITY_INFO, "AutoTrim cancelled");
+    // restore original trims
 }
 
-void Copter::auto_trim()
+void RC_Channels_Copter::auto_trim_run()
 {
-    if (auto_trim_counter > 0) {
-        if (copter.flightmode != &copter.mode_stabilize ||
-            !copter.motors->armed()) {
-            auto_trim_cancel();
+        if (!auto_trim.running) {
             return;
         }
 
-        // flash the leds
-        AP_Notify::flags.save_trim = true;
-
-        if (!auto_trim_started) {
-            if (ap.land_complete) {
-                // haven't taken off yet
-                return;
-            }
-            auto_trim_started = true;
+        // only trim in certain modes:
+        if (copter.flightmode != &copter.mode_stabilize &&
+            copter.flightmode != &copter.mode_althold) {
+            auto_trim_cancel();
+            return;
         }
 
-        if (ap.land_complete) {
-            // landed again.
+        // must be started and stopped mid-air:
```

Comment:
@peterbarker so this cannot be activated before takeoff? I assume land_complete_maybe is true while disarmed and thru arm until liftoff??